File: src/contest/misc/RootSolver.java
package contest.misc;

import java.util.*;
import java.io.*;
import java.math.BigDecimal;
import java.math.RoundingMode;

public class RootSolver {
	static final BigDecimal EPS = new BigDecimal(1e-10);
	static final BigDecimal INF = new BigDecimal(1e20);
	static final int PRECISION = 15;
	
	static BufferedReader br;
	static PrintWriter out;
	static StringTokenizer st;

	public static void main (String[] args) throws IOException {
		br = new BufferedReader(new InputStreamReader(System.in));
		out = new PrintWriter(new OutputStreamWriter(System.out));
		//br = new BufferedReader(new FileReader("in.txt"));
		//out = new PrintWriter(new FileWriter("out.txt"));

		int N = readInt();
		ArrayList<Monomial> polynomial = new ArrayList<Monomial>();
		
		for (int i = 0; i < N; i++)
			polynomial.add(new Monomial(next(), readInt()));
		
		ArrayList<BigDecimal> res = findAllRoots(polynomial);
		
		if (res.size() == 0) {
			out.println("NO REAL ROOTS");
		} else {
			for (BigDecimal d : res) {
				out.println(d.setScale(6, RoundingMode.HALF_UP).toString());
			}
		}
		
		out.close();
	}

	static class Monomial {
		BigDecimal coefficient;
		int exponent;
		

		Monomial (BigDecimal coefficient, int exponent) {
			this.coefficient = coefficient.setScale(PRECISION, RoundingMode.HALF_UP);
			this.exponent = exponent;
		}
		
		Monomial (String coefficient, int exponent) {
			this.coefficient = new BigDecimal(coefficient).setScale(PRECISION, RoundingMode.HALF_UP);
			this.exponent = exponent;
		}
		
		BigDecimal evaluate (BigDecimal x) {
			return coefficient.multiply(x.pow(exponent));
		}
		
		public String toString () {
			return String.format("%sx^%d", coefficient.toString(),  exponent);
		}
	}
	
	static BigDecimal evaluate (ArrayList<Monomial> polynomial, BigDecimal x) {
		BigDecimal ret = BigDecimal.ZERO.setScale(PRECISION, RoundingMode.HALF_UP);
		for (Monomial m : polynomial)
			ret = ret.add(m.evaluate(x));
		return ret;
	}
	
	static BigDecimal findRoot (ArrayList<Monomial> polynomial, BigDecimal x1, BigDecimal x2) {
		BigDecimal y1 = evaluate(polynomial, x1), y2 = evaluate(polynomial, x2);
		boolean neg1 = y1.compareTo(BigDecimal.ZERO) < 0, neg2 = y2.compareTo(BigDecimal.ZERO) < 0;
		
		if (y1.abs().compareTo(EPS) <= 0)
			return x1;
		if (y2.abs().compareTo(EPS) <= 0)
			return null;
		
		if (neg1 == neg2)
			return null;
		
		x2 = x2.setScale(PRECISION, RoundingMode.HALF_UP);
		x1 = x1.setScale(PRECISION, RoundingMode.HALF_UP);
		
		while (x2.subtract(x1).compareTo(EPS) > 0) {
			BigDecimal x = BigDecimal.ZERO.setScale(PRECISION, RoundingMode.HALF_UP);
			x = x.add(x2.add(x1).divide(new BigDecimal("2")));
			if ((evaluate(polynomial, x).compareTo(BigDecimal.ZERO) < 0) == neg1)
				x1 = x;
			else
				x2 = x;
		}
		return x1;
	}
	
	static ArrayList<BigDecimal> findAllRoots (ArrayList<Monomial> polynomial) {
		ArrayList<Monomial> diff = new ArrayList<Monomial>();
		ArrayList<BigDecimal> ret = new ArrayList<BigDecimal>();
		
		for (Monomial m : polynomial)
			if (m.exponent > 0)
				diff.add(new Monomial(m.coefficient.multiply(new BigDecimal(Integer.toString(m.exponent)).setScale(PRECISION, RoundingMode.HALF_UP)), m.exponent - 1));
		
		if (diff.isEmpty())
			return ret;
		
		ArrayList<BigDecimal> diffRoots = findAllRoots(diff);
		diffRoots.add(0, INF.negate());
		diffRoots.add(INF);
		
		
		for (int i = 0; i < diffRoots.size() - 1; i++) {
			BigDecimal root = findRoot(polynomial, diffRoots.get(i), diffRoots.get(i + 1));
			if (root == null)
				continue;
			if (ret.isEmpty() || ret.get(ret.size() - 1).subtract(root).abs().compareTo(EPS) > 0)
				ret.add(root);
		}
		
		return ret;
	}
	
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}

	static long readLong () throws IOException {
		return Long.parseLong(next());
	}

	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}

	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}

	static char readCharacter () throws IOException {
		return next().charAt(0);
	}

	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: src/contest/misc/RootSolver.cc
#include <bits/stdc++.h>
#include <limits>

#define mp make_pair
#define pb push_back
#define INF 1e20
#define MOD 1000000007
#define rint(x) scanf("%d", &(x))
#define rlong(x) scanf("%lld", &(x))
#define SIZE 500100
#define EPS 1e-9
#define NaN numeric_limits<ld>::quiet_NaN();
#define l(x) x << 1
#define r(x) x << 1 | 1
#define m(x, y) (x + y) / 2

using namespace std;

typedef long long ll;
typedef pair<int, int> pi;
typedef pair<ll, ll> pll;
typedef long double ld;

int N;

vector<pair<ld, ld>> polynomial;

ld evaluate(vector<pair<ld, ld>> polynomial, ld x) {
  ld ret = 0;
  for (auto monomial : polynomial)
    ret += monomial.first * powl(x, monomial.second);
  return ret;
}

ld findRoot(vector<pair<ld, ld>> polynomial, ld x1, ld x2) {
  ld y1 = evaluate(polynomial, x1), y2 = evaluate(polynomial, x2);
  bool neg1 = y1 < 0, neg2 = y2 < 0;

  if (abs(y1) <= 0)
    return x1;

  if (abs(y2) <= 0 || neg1 == neg2)
    return NaN;

  while (x2 - x1 > EPS) {
    ld x = (x2 + x1) / 2;
    if ((evaluate(polynomial, x) < 0) == neg1)
      x1 = x;
    else
      x2 = x;
  }
  return x1;
}

vector<ld> findAllRoots(vector<pair<ld, ld>> polynomial) {
  vector<pair<ld, ld>> diff;
  vector<ld> ret;

  for (auto monomial : polynomial)
    if (monomial.second > 0)
      diff.pb({monomial.first * monomial.second, monomial.second - 1});

  if (diff.empty())
    return ret;

  vector<ld> diffRoots = findAllRoots(diff);
  diffRoots.insert(diffRoots.begin(), -INF);
  diffRoots.pb(INF);

  for (int i = 0; i < (int)diffRoots.size() - 1; i++) {
    ld root = findRoot(polynomial, diffRoots[i], diffRoots[i + 1]);
    if (root != root)
      continue;
    if (ret.empty() || root != ret.back())
      ret.pb(root);
  }

  return ret;
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  // freopen("in.txt", "r", stdin);
  // freopen("out.txt", "w", stdout);

  cin >> N;

  for (int i = 0; i < N; i++) {
    ld coefficient, power;
    cin >> coefficient >> power;
    polynomial.pb({coefficient, power});
  }

  vector<ld> res = findAllRoots(polynomial);

  if (res.size() == 0)
    cout << "NO REAL ROOTS\n";
  else {
    cout.precision(9);
    for (ld d : res)
      cout << fixed << d << "\n";
  }

  return 0;
}
